From cdb68d4316e25b7898306b41e3fe100cca454c89 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Wed, 14 May 2025 14:10:27 +0200
Subject: [PATCH] mailbox: stm32-ippc: Add wake up of the system

Two features must be supported:
- Allow IPCC to be a wake-up source.
- Enable the use of IPCC during suspend and resume to configure RCC
  through the SCMI.

The IRQF_NO_SUSPEND allows to use the SCMI but in such case we need to
manually wake up the system in suspend to idle.

Change-Id: Ic0a0a94ed9ce87014020ed231b571a80eb04e67f
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/460853
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 drivers/mailbox/stm32-ipcc.c | 25 +++++++++++++++++++++----
 1 file changed, 21 insertions(+), 4 deletions(-)

--- a/drivers/mailbox/stm32-ipcc.c
+++ b/drivers/mailbox/stm32-ipcc.c
@@ -14,6 +14,7 @@
 #include <linux/of.h>
 #include <linux/platform_device.h>
 #include <linux/pm_wakeirq.h>
+#include <linux/suspend.h>
 #include <linux/workqueue.h>
 
 #include "mailbox.h"
@@ -85,6 +86,7 @@ struct stm32_ipcc {
 	u32 xcr;
 	u32 xmr;
 	u32 sec_mode;
+	bool sys_wkup;
 };
 
 static inline void stm32_ipcc_set_bits(spinlock_t *lock, void __iomem *reg,
@@ -140,6 +142,11 @@ static irqreturn_t stm32_ipcc_rx_irq(int
 
 		dev_dbg(dev, "%s: chan:%d rx\n", __func__, chan);
 
+		if (ipcc->sys_wkup) {
+			pm_system_wakeup();
+			ipcc->sys_wkup = false;
+		}
+
 		chnl = &ipcc->chnl[chan];
 
 		/*
@@ -498,12 +505,17 @@ static int stm32_ipcc_remove(struct plat
 static int stm32_ipcc_suspend(struct device *dev)
 {
 	struct stm32_ipcc *ipcc = dev_get_drvdata(dev);
+	int ret;
 
 	ipcc->xmr = readl_relaxed(ipcc->reg_proc + IPCC_XMR);
 	ipcc->xcr = readl_relaxed(ipcc->reg_proc + IPCC_XCR);
 
-	if (device_may_wakeup(dev))
-		return enable_irq_wake(ipcc->irqs[IPCC_IRQ_RX]);
+	if (device_may_wakeup(dev)) {
+		ret = enable_irq_wake(ipcc->irqs[IPCC_IRQ_RX]);
+		if (ret)
+			return ret;
+		ipcc->sys_wkup = true;
+	}
 
 	return 0;
 }
@@ -511,12 +523,17 @@ static int stm32_ipcc_suspend(struct dev
 static int stm32_ipcc_resume(struct device *dev)
 {
 	struct stm32_ipcc *ipcc = dev_get_drvdata(dev);
+	int ret;
 
 	writel_relaxed(ipcc->xmr, ipcc->reg_proc + IPCC_XMR);
 	writel_relaxed(ipcc->xcr, ipcc->reg_proc + IPCC_XCR);
 
-	if (device_may_wakeup(dev))
-		return disable_irq_wake(ipcc->irqs[IPCC_IRQ_RX]);
+	if (device_may_wakeup(dev)) {
+		ret = disable_irq_wake(ipcc->irqs[IPCC_IRQ_RX]);
+		if (ret)
+			return ret;
+		ipcc->sys_wkup = false;
+	}
 
 	return 0;
 }
